File: baekjoon/stage20/10828_스택.py
import sys

class Stack:
    def __init__(self):
        self.top = -1
        self.arr = []
    
    def get_size(self):
        return self.top + 1
    
    def is_empty(self):
        if self.top == -1:
            return True
        return False
    
    def is_full(self):
        if self.get_size() >= 10000:
            return True
        return False
        
    def push(self, n):
        if self.is_full():
            return
        self.top += 1
        self.arr.append(n)
        
    def pop(self):
        if self.is_empty():
            return -1
        result = self.arr[self.top]
        del self.arr[self.top]
        self.top -= 1
        return result
        
    def get_top(self):
        if self.is_empty():
            return -1
        return self.arr[self.top]

if __name__ == "__main__":
    n = int(sys.stdin.readline())
    stack = Stack()
    for i in range(0, n):
        commands = sys.stdin.readline().split()
        if commands[0] == "push":
            stack.push(int(commands[1]))
        elif commands[0] == "pop":
            print(stack.pop())
        elif commands[0] == "size":
            print(stack.get_size())
        elif commands[0] == "empty":
            if stack.is_empty():
                print(1)
            else:
                print(0)
        elif commands[0] == "top":
            print(stack.get_top())

File: problem-solving/baekjoon/10828_스택.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	var n int
	fmt.Fscanln(reader, &n)
	var s = NewStack()
	s.Init()
	for i := 0; i < n; i++ {
		var command string
		command, _ = reader.ReadString('\n')
		command = strings.ReplaceAll(command, "\n", "")
		commands := strings.Split(command, " ")
		if commands[0] == "push" {
			num, _ := strconv.Atoi(commands[1])
			s.Push(num)
		} else if commands[0] == "pop" {
			fmt.Fprintln(writer, s.Pop())
		} else if commands[0] == "size" {
			fmt.Fprintln(writer, s.GetSize())
		} else if commands[0] == "empty" {
			if s.IsEmpty() {
				fmt.Fprintln(writer, 1)
			} else {
				fmt.Fprintln(writer, 0)
			}
		} else if commands[0] == "top" {
			fmt.Fprintln(writer, s.GetTop())
		}
	}
}

type Stack interface {
	GetSize() int
	IsEmpty() bool
	Push(int)
	Pop() int
	GetTop() int
	Init()
	IsFull() bool
}

type stack struct {
	arr [10000]int
	top int
}

func NewStack() Stack {
	return &stack{}
}

func (s *stack) Init() {
	s.top = -1
}

func (s *stack) GetSize() int {
	return s.top + 1
}
func (s *stack) IsEmpty() bool {
	if s.top == -1 {
		return true
	}
	return false
}

func (s *stack) IsFull() bool {
	if s.GetSize() >= 10000 {
		return true
	}
	return false
}

func (s *stack) Push(n int) {
	if s.IsFull() {
		return
	}
	s.top++
	s.arr[s.top] = n
}

func (s *stack) Pop() int {
	if s.IsEmpty() {
		return -1
	}
	result := s.arr[s.top]
	s.top--
	return result
}

func (s *stack) GetTop() int {
	if s.IsEmpty() {
		return -1
	}
	return s.arr[s.top]
}
